pkg/api/message/service.go

```diff
@@ -117,19 +122,104 @@ func (s *Service) SubscribeEnvelopes(
 	}
 }
 
+// Pulls from DB and sends to client, updating the query's last seen cursor, until
+// the stream has caught up to the latest in the database.
+func (s *Service) catchUpFromCursor(
+	stream message_api.ReplicationApi_SubscribeEnvelopesServer,
+	query *message_api.EnvelopesQuery,
+	logger *zap.Logger,
+) error {
+	log := logger.With(zap.String("stage", "catchUpFromCursor"))
+	if query.GetLastSeen() == nil {
+		log.Debug("Skipping catch up")
+		// Requester only wants new envelopes
+		return nil
+	}
+
+	cursor := query.LastSeen.GetNodeIdToSequenceId()
+	// GRPC does not distinguish between empty map and nil
+	if cursor == nil {
+		cursor = make(map[uint32]uint64)
+		query.LastSeen.NodeIdToSequenceId = cursor
+	}
+
+	log.Debug("Catching up from cursor", zap.Any("cursor", cursor))
+	for {
+		rows, err := s.fetchEnvelopes(stream.Context(), query, maxRequestedRows)
+		log.Debug("Fetched envelopes", zap.Any("rows", rows))
+		if err != nil {
+			return err
+		}
+		envs := make([]*envelopes.OriginatorEnvelope, 0, len(rows))
+		for _, r := range rows {
+			env, err := envelopes.NewOriginatorEnvelopeFromBytes(r.OriginatorEnvelope)
+			if err != nil {
+				// We expect to have already validated the envelope when it was inserted
+				logger.Error("could not unmarshal originator envelope", zap.Error(err))
+				continue
+			}
+			envs = append(envs, env)
+		}
+		err = s.sendEnvelopes(stream, query, envs)
+		if err != nil {
+			return status.Errorf(codes.Internal, "error sending envelopes: %v", err)
+		}
+		if len(rows) < int(maxRequestedRows) {
+			// There were no more envelopes in DB at time of fetch
+			break
+		}
+		// TODO(rich): Determine default rate limits and set query interval to match
+		time.Sleep(50 * time.Millisecond)
+	}
+	return nil
+}
+
+func (s *Service) sendEnvelopes(
+	stream message_api.ReplicationApi_SubscribeEnvelopesServer,
+	query *message_api.EnvelopesQuery,
+	envs []*envelopes.OriginatorEnvelope,
+) error {
+	cursor := query.GetLastSeen().GetNodeIdToSequenceId()
+	if cursor == nil {
+		cursor = make(map[uint32]uint64)
+		query.LastSeen = &envelopesProto.VectorClock{
+			NodeIdToSequenceId: cursor,
+		}
+	}
+	for _, env := range envs {
+		if cursor[uint32(env.OriginatorNodeID())] >= env.OriginatorSequenceID() {
+			continue
+		}
+
+		// TODO(rich): Either batch send envelopes, or modify stream proto to
```

Comment:
WDYT?

I like the idea of batching, since it lets clients update their bookkeeping of the lastSeen cursor once per batch instead of once per envelope.

Reply:
Sounds good - will add to next PR